File: go/runner/results.go
package runner

import (
	"encoding/json"
	"os"
	"path/filepath"

	"github.com/varar-dev/varar/go/core"
)

// Persists run results for the language server (ADR 0014) — the shell half of
// the contract the core builds the payload for. Writes
// <root>/.varar/<oathPath>.json, which the (language-neutral) LSP reads to turn
// a failure into an editor diagnostic.
//
// Lives in the runner so every adapter in this port feeds the same collector
// and cannot drift from the TypeScript reporter this is a port of.

// ResultFilePath is <root>/.varar/<oathPath>.json — the file the LSP watches.
func ResultFilePath(root, oathPath string) string {
	return filepath.Join(root, ".varar", filepath.FromSlash(oathPath)+".json")
}

// WriteOathResults writes one oath's results: 2-space indent plus a trailing
// newline, matching JSON.stringify(results, null, 2) in the TypeScript port.
// SetEscapeHTML(false) is what keeps <, > and & raw the way JSON.stringify
// leaves them; the struct tags in core/result.go fix the field names and order.
func WriteOathResults(root string, results core.OathResults) (string, error) {
	out := ResultFilePath(root, results.OathPath)
	if err := os.MkdirAll(filepath.Dir(out), 0o755); err != nil {
		return "", err
	}
	var buf []byte
	encoded, err := marshalIndentNoEscape(results)
	if err != nil {
		return "", err
	}
	buf = append(encoded, '\n')
	if err := os.WriteFile(out, buf, 0o644); err != nil {
		return "", err
	}
	return out, nil
}

func marshalIndentNoEscape(value any) ([]byte, error) {
	var buf []byte
	writer := &sliceWriter{buf: &buf}
	encoder := json.NewEncoder(writer)
	encoder.SetEscapeHTML(false)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(value); err != nil {
		return nil, err
	}
	// Encode appends its own newline; WriteOathResults adds the one we want.
	return buf[:len(buf)-1], nil
}

type sliceWriter struct{ buf *[]byte }

func (w *sliceWriter) Write(p []byte) (int, error) {
	*w.buf = append(*w.buf, p...)
	return len(p), nil
}

// Results accumulates each oath's example results across a run, then writes
// them. `go test` reports test by test and has no end-of-run hook of its own,
// so the adapter flushes once every subtest has run — the first moment an
// oath's examples are all in. Passing oaths are written too: a stale file would
// keep a diagnostic on screen that the run has just cleared.
type Results struct {
	order    []string
	sources  map[string]string
	examples map[string][]core.ExampleResult
}

// NewResults is an empty collector.
func NewResults() *Results {
	return &Results{sources: map[string]string{}, examples: map[string][]core.ExampleResult{}}
}

// Record accumulates one example's outcome.
func (r *Results) Record(oathPath, source string, result core.ExampleResult) {
	if _, seen := r.sources[oathPath]; !seen {
		r.order = append(r.order, oathPath)
	}
	r.sources[oathPath] = source
	r.examples[oathPath] = append(r.examples[oathPath], result)
}

// FlushAll writes every oath held, and forgets them. Write errors are ignored
// on purpose: a read-only workspace must not fail a test run whose results are
// otherwise fine — the editor simply shows nothing for it.
func (r *Results) FlushAll(root string) {
	for _, oathPath := range r.order {
		examples := r.examples[oathPath]
		if len(examples) == 0 {
			continue
		}
		_, _ = WriteOathResults(root, core.OathResults{
			Version:    1,
			OathPath:   oathPath,
			SourceHash: core.HashSource(r.sources[oathPath]),
			Examples:   examples,
		})
	}
	r.order = nil
	r.sources = map[string]string{}
	r.examples = map[string][]core.ExampleResult{}
}

File: rust/runner/src/results.rs
//! Persists run results for the language server (ADR 0014) — the shell half of
//! the contract the core builds the payload for. Writes
//! `<root>/.varar/<oath_path>.json`, which the (language-neutral) LSP reads to
//! turn a failure into an editor diagnostic.
//!
//! Lives in the runner so every adapter in this port feeds the same collector
//! and cannot drift from the TypeScript reporter this is a port of.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use varar_core::hash::hash_source;
use varar_core::result::{ExampleResult, OathResults, ReferencedDocument, to_wire_json};

/// `<root>/.varar/<oath_path>.json` — the file the LSP watches.
pub fn result_file_path(root: &Path, oath_path: &str) -> PathBuf {
    root.join(".varar").join(format!("{oath_path}.json"))
}

/// Writes one oath's results: 2-space indent plus a trailing newline, matching
/// `JSON.stringify(results, null, 2)` in the TypeScript port.
pub fn write_oath_results(root: &Path, results: &OathResults) -> std::io::Result<PathBuf> {
    let out = result_file_path(root, &results.oath_path);
    if let Some(parent) = out.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&out, format!("{}\n", to_wire_json(results)))?;
    Ok(out)
}

/// Sorts examples into document order.
///
/// A test framework reports examples in ITS order — `cargo test` runs them in
/// parallel, so the order results are recorded in is not the order they appear
/// in the oath. The file is a cross-port contract read by tools that diff runs,
/// so it is written in document order everywhere. The name breaks ties for
/// examples sharing a line.
fn document_order(examples: &mut [ExampleResult]) {
    examples.sort_by(|a, b| {
        let line = |e: &ExampleResult| e.lines.first().copied().unwrap_or(0);
        line(a).cmp(&line(b)).then_with(|| a.name.cmp(&b.name))
    });
}

/// Accumulates each oath's example results across a run, then writes them.
///
/// `cargo test` reports one test at a time and has no end-of-run hook of its
/// own, so the harness flushes this once `libtest_mimic::run` returns — the
/// first moment an oath's examples are all in. Passing oaths are written too: a
/// stale file would keep a diagnostic on screen that the run has just cleared.
#[derive(Default)]
pub struct Results {
    sources: BTreeMap<String, String>,
    examples: BTreeMap<String, Vec<ExampleResult>>,
    /// Per oath: the other documents its steps were spliced in from (ADR 0016),
    /// as path → source.
    documents: BTreeMap<String, BTreeMap<String, String>>,
}

impl Results {
    pub fn new() -> Results {
        Results::default()
    }

    /// Accumulates one example's outcome. `referenced_sources` carries the OTHER
    /// documents this oath's steps were spliced in from (ADR 0016), as
    /// path → source; their hashes go in the payload so a consumer can tell a
    /// stale failure from a live one.
    pub fn record(
        &mut self,
        oath_path: &str,
        source: &str,
        result: ExampleResult,
        referenced_sources: &BTreeMap<String, String>,
    ) {
        self.sources
            .entry(oath_path.to_string())
            .or_insert_with(|| source.to_string());
        self.examples
            .entry(oath_path.to_string())
            .or_default()
            .push(result);
        if !referenced_sources.is_empty() {
            self.documents
                .entry(oath_path.to_string())
                .or_default()
                .extend(
                    referenced_sources
                        .iter()
                        .map(|(k, v)| (k.clone(), v.clone())),
                );
        }
    }

    /// Writes every oath held, and forgets them. Errors are ignored on purpose:
    /// a read-only or missing workspace must not fail a test run whose results
    /// are otherwise fine — the editor simply shows nothing for it.
    pub fn flush_all(&mut self, root: &Path) {
        for (oath_path, mut examples) in std::mem::take(&mut self.examples) {
            let Some(source) = self.sources.get(&oath_path) else {
                continue;
            };
            document_order(&mut examples);
            let documents = self
                .documents
                .get(&oath_path)
                .map(|docs| {
                    docs.iter()
                        .map(|(path, text)| ReferencedDocument {
                            path: path.clone(),
                            source_hash: hash_source(text),
                        })
                        .collect()
                })
                .unwrap_or_default();
            let results = OathResults {
                version: 2,
                oath_path: oath_path.clone(),
                source_hash: hash_source(source),
                documents,
                examples,
            };
            let _ = write_oath_results(root, &results);
        }
    }
}
